### PR TITLE
Funding rate taker UI

### LINKED ISSUE #244: Funding rate
The maker has to be able to define interest on the CFD, i.e. a percentage of the taker's amount that will always go to the maker regardless of the price change.

### PR DESCRIPTION
**Pull request by da-kami**
Resolves #244 

TODO list until this can get merged:

- [x] Taker UI display costs when opening and running costs on CFD tile
- [x] Business logic (maker&taker) to pass in opening fee into contract setup
- [x] Business logic (maker&taker) to handle recurring fee on every rollover
- [x] Maker API for creating `Offer` to allow configuration of fees

### CODE REVIEW

File: daemon/src/projection.rs
Comment (da-kami) [inline daemon/src/projection.rs]: We should use `round` here instead of `period`. I think we agreed on round, introducing `period` is yet another term.

File: daemon/src/projection.rs
Comment (da-kami) [inline daemon/src/projection.rs]: I think we might additionally need to define the `funding_round_interval` (or similar name), that expresses how often  `funding_fee_per_period` is collected. 

This should be equal to `settlement_interval`, but it is currently not, until we resolve discussion: https://github.com/itchysats/itchysats/discussions/1068

Comment (da-kami) [reply daemon/src/projection.rs]: added.

File: daemon/src/projection.rs
Comment (da-kami) [reply daemon/src/projection.rs]: renamed.

**Comment by da-kami** [comment]
First version of the taker UI changes done:

This includes:

- The pulsating price is now the maker's ask price (because that is relevant in the context of opening a position, not the index price)
- The index price is currently not displayed anymore, because we did not think it is needed. The fact that that price is used for oracle attestations should be mentioned somewhere - but for collab close (which is the go-to path) it is actually *not* used.
- Basic display of margin and fee as table before pressing `Long`
- Open confirm modal adds a row that states that it is perpetual
- We just added the perpetual cost information to the CFD tile for now (stacke on top of the other values)

This work also includes a tile overhaul:

- Change the background and tile layout and better alignment of components
- Close button is in the right bottom corner, transactions are displayed as table.
- Color `P/L percent` and `Payout` according to profit, loss or zero.
- Distinguish between open/close display.
- Display `Lock + Payout` transactions by default
- `Quantity` was renamed to `Contracts` (to go away from the notion of "thinking in USD")

**Comment by da-kami** [comment]
Current state:

image

image

still way to cluttered with information. But it's gonna be hard to hide it without breaking the layout. Not sure we should invest much more time into this version. Probably better to design it new :)

**Comment by da-kami** [comment]
image

**Comment by da-kami** [comment]
Note: After rebasing the displayed transaction history is a bit messed up, I'll look into that and push a fixup.
The API is stable now @klochowicz

**Comment by klochowicz** [comment]
> Note: After rebasing the displayed transaction history is a bit messed up, I'll look into that and push a fixup. The API is stable now @klochowicz

the API got extracted from this PR into: #1129

**Comment by klochowicz** [comment]
> Resolves #244
> 
> So far still very much work in progress :)
> 
> TODO list until this can get merged:
> 
>     * [x]  Taker UI display costs when opening and running costs on CFD tile
> 
>     * [x]  Business logic (maker&taker) to pass in opening fee into contract setup
> 
>     * [x]  Business logic (maker&taker) to handle recurring fee on every rollover
> 
>     * [x]  Maker API for creating `Offer` to allow configuration of fees
> 
> 
> (note: we could merge this with dummy values as well, but probably should not :)

I've updated the TODO items here for the ones that got already merged. There are still a few small issues that we need to resolve for the full functionality though:
- [ ] #1131 
- [ ]  #1114 
- [ ]  #1125

**Comment by da-kami** [comment]
Once #1121 and #1129 merge I'll rebase again and then we can see to get this one in

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs:2260]: Note: This change was not strictly needed, but it is actually more correct, because we do test the `taker` with these... We should probably be more explicit once we create a more elaborate test framework...

File: daemon/tests/happy_path.rs
Comment (thomaseizinger) [inline daemon/tests/happy_path.rs:69]: Oh my god ...

File: daemon/src/monitor.rs
Comment (thomaseizinger) [inline daemon/src/monitor.rs]: I am fixing this here: https://github.com/itchysats/itchysats/pull/1143

File: taker-frontend/src/App.tsx
Comment (thomaseizinger) [inline taker-frontend/src/App.tsx]: Does `closed` here refer to the state of the CFD?

Seems wrong to declare this at the top here. Why can't each CFD tile figure this out by themselves by looking at the state?

File: taker-frontend/src/components/Trade.tsx
Comment (thomaseizinger) [inline taker-frontend/src/components/Trade.tsx:104]: At this point it would be good to introduce components for this stuff and add tests :)

File: taker-frontend/src/components/Trade.tsx
Comment (thomaseizinger) [inline taker-frontend/src/components/Trade.tsx]: Why aren't we doing this in the backend?

**Review by thomaseizinger — COMMENTED** [review]
Not sure what is going on with this `isClosed` property, otherwise LGTM.

File: daemon/tests/happy_path.rs
Comment (da-kami) [reply daemon/tests/happy_path.rs:69]: yeah...

File: daemon/src/monitor.rs
Comment (da-kami) [reply daemon/src/monitor.rs]: Ok, I'll remove this then

File: taker-frontend/src/App.tsx
Comment (da-kami) [reply taker-frontend/src/App.tsx]: We split the CFDs into two separate lists and I wanted to adhere to that. This just keeps the logic of "splitting it" in one place, otherwise I have to reason twice - one time outside of `Histroy` and one time inside...

File: taker-frontend/src/components/Trade.tsx
Comment (da-kami) [reply taker-frontend/src/components/Trade.tsx]: oh, we actually are. That is legacy that should be changed!

**Comment by da-kami** [comment]
TODO: 

- [x] Multiply opening fee by parcel size (to have proper fee)
- [x] Remove `next_funding_event` from the `CfdOrder` and just take the next full hour in the UI

File: taker-frontend/src/App.tsx
Comment (da-kami) [reply taker-frontend/src/App.tsx]: I changed it into something more specific

File: taker-frontend/src/components/Trade.tsx
Comment (bonomat) [inline taker-frontend/src/components/Trade.tsx]: nit: wow... imho this now looks gigantic 😬

File: taker-frontend/src/components/Trade.tsx
Comment (bonomat) [inline taker-frontend/src/components/Trade.tsx]: Nit: The line-break here removes the whitespace between `₿{btcToOpenPosition}` and `will`

image

File: taker-frontend/src/components/Trade.tsx
Comment (bonomat) [inline taker-frontend/src/components/Trade.tsx:365]: That tooltop bothered me for a long time 😬

File: taker-frontend/src/components/History.tsx
Comment (bonomat) [inline taker-frontend/src/components/History.tsx:152]: I'm not convinced it is a good idea to prominently tell the user how much he is losing through fees... 

Also, note that this value might be positive or negative depending on your position. Would it be clear to you what it means if `Total fees = -0.01234` ?

File: taker-frontend/src/components/History.tsx
Comment (bonomat) [inline taker-frontend/src/components/History.tsx:128]: I assume the `cfd.profit_btc` includes the total fees already. If so, what do you think of adding a tooltip explaining this ?

File: taker-frontend/src/components/History.tsx
Comment (bonomat) [inline taker-frontend/src/components/History.tsx:133]: Question: is the margin constant? 

I think, the _current_ `margin` should be _initial_ `margin` minus `fees_paid`. 
Meaning, the longer your position is running, the more fees you have paid, the lower your margin. 
That being said, we could display this to the user ala `Margin = initial funding - fees paid = 0.003 - 0.0001 = ...` (provided we have this data available.

**Review by bonomat — APPROVED** [review]
Well done. I like how the tiles are evolving. 

As mentioned on a call: we know already that this layout won't work once the user can go long and short because we will need to display 2 exchange rates. 
I think the approach of having only 1 exchange rate is nice, but maybe we should just remove the `Short` button for now.

File: taker-frontend/src/App.tsx
Comment (thomaseizinger) [reply taker-frontend/src/App.tsx]: > We split the CFDs into two separate lists and I wanted to adhere to that. This just keeps the logic of "splitting it" in one place, otherwise I have to reason twice - one time outside of `Histroy` and one time inside...

You always reason about the same thing though which is the state of the CFD? You can easily keep this DRY by extracting that in a function.

What we currently have is pretty confusing because in order to understand what happens, you need to know that the reason this is hard-coded to `false` here is because _a different_ piece of code groups these CFDs into closed and non-closed.

So now you have semantic coupling that is not enforced by the code. If the condition on how you split these CFDs changes without updating this flag, the code inside the tiles will behave incorrectly. The CFD tile already knows its state, it should reason about that locally and not require it to be told externally something that it already knows.

File: taker-frontend/src/components/History.tsx
Comment (da-kami) [reply taker-frontend/src/components/History.tsx:133]: Given that we cannot add to lock at the moment our margin is not constant but changes with every funding event (i.e. we never add "fresh" money to keep the margin constant or reduce it).

My takeaway from these comments is, that the relationship of margin and fees is currently not clear enough. I will see to display the margin as current margin (i.e. initial margin - fees) and add a tooltip that shows initial margin and costs. That way we also don't show the costs in the tile itself. (I am not sure we have the initial margin available, but if we don't we should change that...)

File: taker-frontend/src/components/History.tsx
Comment (da-kami) [reply taker-frontend/src/components/History.tsx:128]: Thanks for pointing that one out - the profit currently does *not* reflect the costs at the moment. 
This is not totally trivial to adapt (will require additional tests as well and internal changes to keep track of the costs). I will align with @klochowicz on who shall tackle this, until the it is tracked here: https://github.com/itchysats/itchysats/issues/1165

File: taker-frontend/src/components/History.tsx
Comment (da-kami) [reply taker-frontend/src/components/History.tsx:152]: see https://github.com/itchysats/itchysats/pull/1066#discussion_r788236094

File: taker-frontend/src/components/Trade.tsx
Comment (da-kami) [reply taker-frontend/src/components/Trade.tsx]: Yeah - the previous button with the price was actually nice from the looks - it added more to the button. I will see to give this another iteration of changes once this merges.

**Comment by da-kami** [comment]
> Well done. I like how the tiles are evolving.
> 
> As mentioned on a call: we know already that this layout won't work once the user can go long and short because we will need to display 2 exchange rates. I think the approach of having only 1 exchange rate is nice, but maybe we should just remove the `Short` button for now.

I opened https://github.com/itchysats/itchysats/issues/1163 and will tackle it as follow-up

File: taker-frontend/src/components/History.tsx
Comment (da-kami) [reply taker-frontend/src/components/History.tsx:133]: Since this requires a bit more work I will do it as follow-up, see: https://github.com/itchysats/itchysats/issues/1164

File: taker-frontend/src/components/Trade.tsx
Comment (da-kami) [reply taker-frontend/src/components/Trade.tsx]: Reduced it to `xl` - might change this with working on the prices again

**Comment by klochowicz** [comment]
@da-kami you might need to rebase on recent master, as bors is now awaiting status from "block-fixup" job, that recently got introduced (which blocks merging `fixup!` commits)

File: taker-frontend/src/App.tsx
Comment (da-kami) [reply taker-frontend/src/App.tsx]: extracted function `isClosed(cfd: Cfd): boolean` - tried to have the `Cfd` as `class` first so I can have a method, but could not easily achieve the mapping using the class, so I opted for a function that reasons about the `Cfd` instead.

**Pull request merged**
